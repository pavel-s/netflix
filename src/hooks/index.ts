export { useAuth } from './use-auth-user';
export { useAuthUser } from './use-auth-user';
export { useContent } from './use-content';
