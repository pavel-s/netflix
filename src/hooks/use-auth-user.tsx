import { useEffect, useState } from 'react';
import firebase, { auth, TFirebaseUser } from '../lib/firebase.prod';

const USER_STORAGE_KEY = 'authUser';

export const useAuthUser = () => {
  const storedUser = localStorage.getItem(USER_STORAGE_KEY);

  const [user, setUser] = useState<TFirebaseUser | null>(
    storedUser ? JSON.parse(storedUser) : null
  );

  useEffect(() => {
    const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        localStorage.setItem(USER_STORAGE_KEY, JSON.stringify(user));
        setUser(user);
      } else {
        localStorage.removeItem(USER_STORAGE_KEY);
        setUser(null);
      }
    });
    return unsubscribe;
  }, []);

  return user;
};

const signUp = ({
  name,
  email,
  password,
}: {
  name: string;
  email: string;
  password: string;
}): Promise<void> => {
  return auth.createUserWithEmailAndPassword(email, password).then((result) =>
    result.user?.updateProfile({
      displayName: name,
      photoURL: `images/users/${Math.floor(Math.random() * 5) + 1}.png`,
    })
  );
};

const signOut = () => auth.signOut();

const signIn = ({ email, password }: { email: string; password: string }) => {
  return auth.signInWithEmailAndPassword(email, password);
};

const authMethods = { signUp, signOut, signIn };
export const useAuth = () => {
  return authMethods;
};
