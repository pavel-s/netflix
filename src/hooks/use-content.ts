import { useEffect, useState } from 'react';
import firebase from '../lib/firebase.prod';
import { TMediaItem } from '../types';

export const useContent = (type: 'films' | 'series') => {
  const [content, setContent] = useState<TMediaItem[]>([]);
  useEffect(() => {
    firebase
      .firestore()
      .collection(type)
      .get()
      .then((snapshot) => {
        const result = snapshot.docs.map(
          (docObj) =>
            ({
              ...docObj.data(),
              docId: docObj.id,
            } as TMediaItem)
        );
        setContent(result);
      })
      .catch((error) => console.log(error.message));
  }, [type]);
  return content;
};
