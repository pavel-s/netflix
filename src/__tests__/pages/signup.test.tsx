import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router';
import { useAuth } from '../../hooks';
import SignUP from '../../pages/signup';

jest.mock('../../hooks', () => {
  return { useAuth: jest.fn() };
});

describe('<SignUp />', () => {
  it('renders SignUp page', async () => {
    // @ts-ignore
    useAuth.mockReturnValue({
      signUp: () => Promise.resolve('sign up success'),
    });
    const {
      container,
      queryByTestId,
      getByPlaceholderText,
      getByTestId,
      getAllByTestId,
    } = render(<SignUP />, {
      wrapper: MemoryRouter,
    });

    // no error
    expect(queryByTestId('form-error')).toBeFalsy();

    // enter incorrect values
    await act(async () => {
      fireEvent.change(getByPlaceholderText('Name'), {
        target: { value: '' },
      });
      fireEvent.change(getByPlaceholderText('Email or phone number'), {
        target: { value: 'gg' },
      });
      fireEvent.change(getByPlaceholderText('Password'), {
        target: { value: 'gg' },
      });
    });

    await act(async () => {
      fireEvent.click(getByTestId('form-submit'));
    });

    // errors
    expect(getAllByTestId('form-error')).toBeTruthy();

    // enter correct values
    await act(async () => {
      fireEvent.change(getByPlaceholderText('Name'), {
        target: { value: 'Bob' },
      });
      fireEvent.change(getByPlaceholderText('Email or phone number'), {
        target: { value: 'gggrres@fffg.cc' },
      });
      fireEvent.change(getByPlaceholderText('Password'), {
        target: { value: 'ttttt7777qqqq' },
      });
    });

    // submit
    await act(async () => {
      fireEvent.click(getByTestId('form-submit'));
    });

    // new inputs values
    expect(getByPlaceholderText('Name')).toHaveValue('Bob');
    expect(getByPlaceholderText('Email or phone number')).toHaveValue(
      'gggrres@fffg.cc'
    );
    expect(getByPlaceholderText('Password')).toHaveValue('ttttt7777qqqq');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('submit form return error', async () => {
    // @ts-ignore
    useAuth.mockReturnValue({
      signUp: () => Promise.reject({ message: 'sign up reject' }),
    });

    const { container, getByTestId, getByPlaceholderText, getByText } = render(
      <SignUP />,
      {
        wrapper: MemoryRouter,
      }
    );

    // enter correct values
    await act(async () => {
      fireEvent.change(getByPlaceholderText('Name'), {
        target: { value: 'Bob' },
      });
      fireEvent.change(getByPlaceholderText('Email or phone number'), {
        target: { value: 'gggrres@fffg.cc' },
      });
      fireEvent.change(getByPlaceholderText('Password'), {
        target: { value: 'ttttt7777qqqq' },
      });
    });

    // submit
    await act(async () => {
      fireEvent.click(getByTestId('form-submit'));
    });

    // submit error
    expect(getByText('sign up reject')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
