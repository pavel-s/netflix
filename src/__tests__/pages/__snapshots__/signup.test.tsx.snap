// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders SignUp page 1`] = `
<div
  class="header__Background-sc-1qf3jwc-0 dEAuVS"
>
  <div
    class="header__Container-sc-1qf3jwc-1 iJrjfW"
  >
    <img
      alt="Netflix"
      class="header__Logo-sc-1qf3jwc-2 edGsUc"
      src="/images/misc/logo.svg"
    />
  </div>
  <div
    class="form__Container-sc-1e02ltb-0 dTbbCb"
  >
    <h1
      class="form__Title-sc-1e02ltb-2 kkeSEH"
    >
      Sign In
    </h1>
    <form
      class="form__Base-sc-1e02ltb-1 NkBbX"
    >
      
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="name"
        placeholder="Name"
        type="text"
        value="Bob"
      />
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="email"
        placeholder="Email or phone number"
        type="text"
        value="gggrres@fffg.cc"
      />
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="password"
        placeholder="Password"
        type="password"
        value="ttttt7777qqqq"
      />
      <button
        class="form__Submit-sc-1e02ltb-5 fvISnT"
        data-testid="form-submit"
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <p
      class="form__TextBig-sc-1e02ltb-9 qOJIB"
    >
      Already a user?
       
      <a
        class="form__Link-sc-1e02ltb-10 ASZx"
        color="#fff"
        href="/login"
      >
        Sign in
      </a>
    </p>
  </div>
  <div
    class="footer__Container-sc-58hhi-0 gLeidq"
  >
    <div
      class="footer__Inner-sc-58hhi-1 hxpWbN"
    >
      <p
        class="footer__Title-sc-58hhi-2 iYvwUS"
      >
        Questions? Call 
        <a
          href="tel:8-800-100-9668"
        >
          8-800-100-9668
        </a>
      </p>
      <ul
        class="footer__Links-sc-58hhi-3 dVUMzK"
      >
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            FAQ
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Help Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Account
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Media Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Investor Relations
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Jobs
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Ways to Watch
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Terms of Use
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Privacy
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Cookie Preferences
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Corporate Information
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Contact Us
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Speed Test
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Legal Notices
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Netflix Originals
          </a>
        </li>
      </ul>
      <p
        class="footer__Text-sc-58hhi-5 eNZVhR"
      >
        Netflix Russia
      </p>
    </div>
  </div>
</div>
`;

exports[`<SignUp /> submit form return error 1`] = `
<div
  class="header__Background-sc-1qf3jwc-0 dEAuVS"
>
  <div
    class="header__Container-sc-1qf3jwc-1 iJrjfW"
  >
    <img
      alt="Netflix"
      class="header__Logo-sc-1qf3jwc-2 edGsUc"
      src="/images/misc/logo.svg"
    />
  </div>
  <div
    class="form__Container-sc-1e02ltb-0 dTbbCb"
  >
    <h1
      class="form__Title-sc-1e02ltb-2 kkeSEH"
    >
      Sign In
    </h1>
    <form
      class="form__Base-sc-1e02ltb-1 NkBbX"
    >
      <p
        class="form__Error-sc-1e02ltb-4 dprvGA"
        data-testid="form-error"
      >
        sign up reject
      </p>
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="name"
        placeholder="Name"
        type="text"
        value="Bob"
      />
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="email"
        placeholder="Email or phone number"
        type="text"
        value="gggrres@fffg.cc"
      />
      <input
        class="form__Input-sc-1e02ltb-3 gqlhzz"
        id="password"
        placeholder="Password"
        type="password"
        value="ttttt7777qqqq"
      />
      <button
        class="form__Submit-sc-1e02ltb-5 fvISnT"
        data-testid="form-submit"
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <p
      class="form__TextBig-sc-1e02ltb-9 qOJIB"
    >
      Already a user?
       
      <a
        class="form__Link-sc-1e02ltb-10 ASZx"
        color="#fff"
        href="/login"
      >
        Sign in
      </a>
    </p>
  </div>
  <div
    class="footer__Container-sc-58hhi-0 gLeidq"
  >
    <div
      class="footer__Inner-sc-58hhi-1 hxpWbN"
    >
      <p
        class="footer__Title-sc-58hhi-2 iYvwUS"
      >
        Questions? Call 
        <a
          href="tel:8-800-100-9668"
        >
          8-800-100-9668
        </a>
      </p>
      <ul
        class="footer__Links-sc-58hhi-3 dVUMzK"
      >
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            FAQ
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Help Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Account
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Media Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Investor Relations
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Jobs
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Ways to Watch
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Terms of Use
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Privacy
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Cookie Preferences
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Corporate Information
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Contact Us
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Speed Test
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Legal Notices
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Netflix Originals
          </a>
        </li>
      </ul>
      <p
        class="footer__Text-sc-58hhi-5 eNZVhR"
      >
        Netflix Russia
      </p>
    </div>
  </div>
</div>
`;
